### PR TITLE
Metadata: Configure backing field by attribute on skip navigation

### PR DESCRIPTION
**Pull request by smitpatel**

### CODE REVIEW

File: src/EFCore/Metadata/Conventions/BackingFieldConvention.cs [hunk 1]
```diff
@@ -69,11 +72,51 @@ public BackingFieldConvention([NotNull] ProviderConventionSetBuilderDependencies
             IConventionNavigationBuilder navigationBuilder,
             IConventionContext<IConventionNavigationBuilder> context)
         {
-            var navigation = navigationBuilder.Metadata;
-            var field = GetFieldToSet(navigation);
-            if (field != null)
+            if (ConfigurationSource.Convention.Overrides(navigationBuilder.Metadata.GetFieldInfoConfigurationSource()))
```
Comment (smitpatel) [inline src/EFCore/Metadata/Conventions/BackingFieldConvention.cs:75]: I added this condition and put BackingFieldAttributeConvention before this convention so we don't run this if backing field is already set by annotation.

File: src/EFCore/Metadata/Conventions/BackingFieldConvention.cs [hunk 2]
```diff
@@ -57,10 +57,13 @@ public BackingFieldConvention([NotNull] ProviderConventionSetBuilderDependencies
...
-            var field = GetFieldToSet(propertyBuilder.Metadata);
-            if (field != null)
+            if (ConfigurationSource.Convention.Overrides(propertyBuilder.Metadata.GetFieldInfoConfigurationSource()))
             {
-                propertyBuilder.HasField(field);
+                var field = GetFieldToSet(propertyBuilder.Metadata);
+                if (field != null)
+                {
+                    propertyBuilder.HasField(field);
+                }
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Conventions/BackingFieldConvention.cs:66]: You could DRY this up

Comment (smitpatel) [reply src/EFCore/Metadata/Conventions/BackingFieldConvention.cs:66]: how?

Comment (AndriySvyryd) [reply src/EFCore/Metadata/Conventions/BackingFieldConvention.cs:66]: By extracting `IConventionPropertyBaseBuilder`

Comment (smitpatel) [reply src/EFCore/Metadata/Conventions/BackingFieldConvention.cs:66]: HasField API is defined on IConventionProperty/Navigation/SkipNavigationBuilder. Do you want to move those too?

Comment (AndriySvyryd) [reply src/EFCore/Metadata/Conventions/BackingFieldConvention.cs:66]: Yes, that's what I am suggesting

File: src/EFCore/Metadata/Builders/IConventionPropertyBaseBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    ///     </para>
+    ///     <para>
+    ///         This interface is typically used by database providers (and other extensions). It is generally
+    ///         not used in application code.
+    ///     </para>
+    /// </summary>
+    public interface IConventionPropertyBaseBuilder : IConventionAnnotatableBuilder
+    {
+        /// <summary>
+        ///     Gets the property base being configured.
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Builders/IConventionPropertyBaseBuilder.cs:21]:
```suggestion
        ///     Gets the property-like object being configured.
```

**Comment by AndriySvyryd** [comment]
You missed `IConventionServicePropertyBuilder`

**Comment by smitpatel** [comment]
Related test question:
I had to add "new" in actual builders since return type differs. But the ones returning bool are needed for that purpose. But test is failing asking me to add those too. Should I add CanSet methods with new in each of the builder too?

Doesn't this mean that we are writing a lot more code to DRY little code somewhere? :trollface:

**Comment by AndriySvyryd** [comment]
> But the ones returning bool are needed for that purpose. But test is failing asking me to add those too. Should I add CanSet methods with new in each of the builder too?

You should modify the test to look for the methods on base interfaces as well, or just add an exception

> Doesn't this mean that we are writing a lot more code to DRY little code somewhere? :trollface:

Hopefully this refactoring allows to DRY more places

File: src/EFCore/Metadata/Conventions/BackingFieldConvention.cs [hunk 4]
```diff
@@ -57,23 +57,57 @@ public BackingFieldConvention([NotNull] ProviderConventionSetBuilderDependencies
             IConventionPropertyBuilder propertyBuilder,
             IConventionContext<IConventionPropertyBuilder> context)
         {
-            var field = GetFieldToSet(propertyBuilder.Metadata);
-            if (field != null)
-            {
-                propertyBuilder.HasField(field);
-            }
+            DiscoverField(propertyBuilder);
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Conventions/BackingFieldConvention.cs:60]: Might as well run this on service properties too

File: src/EFCore/Metadata/Conventions/SkipNavigationBackingFieldAttributeConvention.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        {
+        }
+
+        /// <inheritdoc />
+        public override void ProcessSkipNavigationAdded(
+            IConventionSkipNavigationBuilder skipNavigationBuilder,
+            BackingFieldAttribute attribute,
+            IConventionContext<IConventionSkipNavigationBuilder> context)
+        {
+            skipNavigationBuilder.HasField(attribute.Name, fromDataAnnotation: true);
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Conventions/SkipNavigationBackingFieldAttributeConvention.cs:31]: Also run this on normal navigations. And run `BackingFieldAttributeConvention` on service properties

Comment (smitpatel) [reply src/EFCore/Metadata/Conventions/SkipNavigationBackingFieldAttributeConvention.cs:31]: Navigation one already does in NavigationBackingFieldAttributeConvention.

Comment (AndriySvyryd) [reply src/EFCore/Metadata/Conventions/SkipNavigationBackingFieldAttributeConvention.cs:31]: Combine them then

Comment (smitpatel) [reply src/EFCore/Metadata/Conventions/SkipNavigationBackingFieldAttributeConvention.cs:31]: Filed https://github.com/dotnet/efcore/issues/21869 for service properties. Bit unsure if we should add those conventions making model building even bigger.

**Pull request merged**